pkg/controller/controlplane/actuator.go

```diff
@@ -0,0 +1,148 @@
+func (a *actuator) Reconcile(
+	ctx context.Context,
+	cp *extensionsv1alpha1.ControlPlane,
+	cluster *extensionscontroller.Cluster,
+) (bool, error) {
+	// Call Reconcile on the composed Actuator
+	requeue, err := a.Actuator.Reconcile(ctx, cp, cluster)
+	if err != nil {
+		return false, err
+	}
+
+	ivector := imagevector.ImageVector()
+	version := cluster.Shoot.Spec.Kubernetes.Version
+
+	// Create shoot chart renderer
+	chartRenderer, err := a.chartRendererFactory.NewChartRendererForShoot(version)
+	if err != nil {
+		return false, fmt.Errorf("could not create chart renderer for shoot '%s': %w", cp.Namespace, err)
+	}
+
+	// extract any subchart named "namespaces" if it exists and pre-process it
+	for _, subc := range a.additionalCharts {
+		if subc.Name == namespacesChartName {
+			// Render each shoot additional components chart and create a managed resource
+			if err := managedresources.RenderChartAndCreate(ctx, cp.Namespace, subc.Name, false, a.client, chartRenderer, subc, nil, ivector, "", version, true, false); err != nil {
+				return false, err
+			}
+		}
+	}
+
+	// Get shoot additional components charts values
+	values, err := a.additionalChartsValuesProvider.GetShootAdditionalChartValues(cp, cluster)
+	if err != nil {
+		return false, err
+	}
+
+	// Render each shoot additional components chart and create a managed resource
+	for _, chart := range a.additionalCharts {
+		// Render each shoot additional components chart and create a managed resource.
+		// We extract any namespace override from the values for the chart.
+		var chartNamespace string
+		if val, ok := values[chart.Name]; !ok {
+			chartNamespace = ""
+		} else if ns, ok := val["namespace"]; !ok {
+			chartNamespace = ""
+		} else if chartNamespace, ok = ns.(string); !ok {
+			return false, fmt.Errorf("shoot additional components namespace was not blank and not a string, instead: %v", val)
+		}
+
+		if err := managedresources.RenderChartAndCreate(ctx, cp.Namespace, chart.Name, false, a.client, chartRenderer, chart, values[chart.Name], ivector, chartNamespace, version, true, false); err != nil {
+			return false, err
+		}
+	}
+	return requeue, nil
+}
+
+// Delete reconciles the given controlplane and cluster, deleting the additional
+// control plane components as needed.
+// Before delegating to the composed Actuator, it ensures that all additional charts have been deleted.
+// It does it in the reverse order of creation
+func (a *actuator) Delete(
+	ctx context.Context,
+	cp *extensionsv1alpha1.ControlPlane,
+	cluster *extensionscontroller.Cluster,
+) error {
+	// delete our additional charts, then the namespace, then the core controlplane
+	// Render each shoot additional components chart and create a managed resource
+	for i := len(a.additionalCharts) - 1; i >= 0; i-- {
+		chart := a.additionalCharts[i]
+		if err := managedresources.Delete(ctx, a.client, cp.Namespace, chart.Name, false); err != nil {
+			return err
+		}
+	}
```

Comment:
Again, the order in which you delete the managed resources doesn't matter, they will be deleted in parallel by GRM. You could change this special for loop to a simple range loop like the one in `Reconcile`.

Reply:
Doesn't that a create issues? I cannot create a resource in a namespace before I create the namespace; I cannot delete the namespace until I delete everything in it.